[docs]: add tutorials: `02-setting-up-azure-functions.md`, `03-understanding-rag.md` and `04-understanding-preparing-language-models.md` (#45)
Azure-Samples/serverless-chat-langchainjs: docs/tutorial/01-introduction.md

# Tutorial - Create a Serverless Chat Application with RAG using LangChain.js and TypeScript

Welcome to the tutorial _Create a Serverless Chat Application with RAG using LangChain.js and TypeScript_.

> This tutorial will guide you through creating a serverless a Chat Application and RAG (Retrieval-Augmented Generation) application using **[LangChain.js](https://js.langchain.com/docs/get_started/introduction)**, **[Azure Functions](https://learn.microsoft.com/azure/azure-functions/)**, **[Azure Cosmos DB for MongoDB vCore](https://learn.microsoft.com/azure/cosmos-db/mongodb/vcore/)**, **[Azure Blob Storage](https://learn.microsoft.com/azure/storage/blobs/)**, and **[Azure Static Web Apps](https://learn.microsoft.com/azure/static-web-apps/)**.

Reviewer: Recommend moving these links to the end of the tutorial as next steps

> The chatbot will be able to answer questions based on a set of enterprise documents uploaded from a fictional company called _Contoso Real Estate_.

Reviewer: ```suggestion
The chatbot you're building can answer questions based on a set of enterprise documents uploaded from a fictional company called _Contoso Real Estate_.
```

Here's an example of the application in action:

![Chat Application with RAG](../../docs/images/demo.gif)

This tutorial will teach you how to build a serverless application using Azure Functions and LangChain.js.

LangChain.js is important because it makes it easy to create language models and integrate them into applications. It also makes it easy to develop AI-driven chatbots. You'll learn how to set up the environment and deploy the application.

> The front end of the application is provided so that you can focus on the backend code and technologies.

Reviewer: can we link to the repo at this point so the user can go check it out if they want

Author: I still need to create the starter branch. I'm just waiting for the final additions that are still going on in the code

## Prerequisites

> You can follow tutorial and run the application using one of the following options:

Reviewer: ```suggestion
You can un the application in the tutorial using one of the following options:
```

- Run the application locally on your machine.
- Run the application using Codespaces.

> > It is highly recommended to use Codespaces for this tutorial. Codespaces is a cloud-based tool that enables you to run development environments without installing any tools on your computer. This way, you can focus on the development process without worrying about the environment setup.

Reviewer: This note seems linked to the Codespaces option

```suggestion
   > It is highly recommended to use Codespaces for this tutorial. Codespaces is a cloud-based tool that enables you 
   to run development environments without installing any tools on your computer. This way, you can focus on the 
   development process without worrying about the environment setup.
```

- [Git](https://git-scm.com/downloads)
- [Azure Developer CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli)

> > If you are a Windows user, you will need to install [PowerShell](https://learn.microsoft.com/powershell/scripting/install/installing-powershell-on-windows?view=powershell-7.4), [Git Bash](https://git-scm.com/downloads) or [WSL2](https://learn.microsoft.com/windows/wsl/install) to run the bash commands.

Reviewer: if this goes through Acrolinx that's used for Docs and Learn, then "you'll" is preferred over "you will"

Author: Dina asked me to use Word's grammar speeling. And Word suggested putting it as : you will. Should I change it to you'll?

The code sample includes sample data to make trying the application quick and easy, but feel free to replace it with your own. You'll use a fictitious company called Contoso Real Estate, and the experience allows its customers to ask support questions about the usage of the company's products. The sample data includes a set of documents that describes the company's terms of service, privacy policy, and support guide.

## Understanding the project architecture

The architecture of the project is shown in the following diagram:

![Chat Application with RAG](../../docs/images/architecture.drawio.png)

To understand the architecture of the project, let's break it down into its individual components:

1. **Web App:**

>    - The user interface for the chatbot is a web application built with **[Lit](https://lit.dev/)** (a library for building web components) and hosted using **[Azure Static Web Apps](https://learn.microsoft.com/azure/static-web-apps/overview)**. It makes a chat interface for users to interact with and ask questions.

Reviewer: ```suggestion
   - The user interface for the chatbot is a web application built with **[Lit](https://lit.dev/)** (a library for building web components) and hosted using **[Azure Static Web Apps](https://learn.microsoft.com/azure/static-web-apps/overview)**. It provides a chat interface for users they can use to ask questions.
```

   - The code is in the `packages/webapp` folder.

2. **Serverless API:**

   - When a user sends a query through the web app, it is sent via HTTP to an API built using Azure Functions.
   - The API uses LangChain.js to process the query.

>    - The API manages the logic of corporate documents and generates answers to chat queries.

Reviewer: ```suggestion
   - The API manages the logic of corporate documents and responds with answers to chat queries.
```

5. **Azure OpenAI Service:**

>    - This service is where the AI Model (a Large Language Model or LLM) is hosted. The model can understand and generating natural language. This is used to embed text chunks or generate answers based on the vector search from the database.

Reviewer: ```suggestion
   - This service is where the AI Model (a Large Language Model or LLM) is hosted. The model can understand and generate natural language. This is used to embed text chunks or generate answers based on the vector search from the database.
```

npm install
```

> 2. To run the project, with only Front-end, execute the following command:

Reviewer: this leaves me wondering as a user if I should run it this way "Front-end only" or with front end and backend (what's the command for that in that case?), when do I know which variant to run?

Author: If the user has already forked the starter branch, they won't be able to run the api. This folder will be deleted. However, the application on the Front-End side will be there, ready to run using the command: `npm run build:webapp`

Now, if you want to run the project as a whole, you'll first need to: provision this application (as there are numerous environment variables for Azure services, which I'm still writing about) and then run the application integrated with Front and Back: `npm run start`
